oauth2client/flask.py

```diff
@@ -0,0 +1,299 @@
+# Copyright 2015 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Utilities for the Flask web framework
+"""
+from __future__ import absolute_import
+import json
+import random
+import string
+from functools import wraps
+
+import httplib2
+from flask import (Blueprint, current_app, redirect, request, session, url_for,
+                   _app_ctx_stack)
+from oauth2client.client import (Credentials, OAuth2WebServerFlow,
+                                 FlowExchangeError)
+from oauth2client import clientsecrets
+
+
+DEFAULT_SCOPES = ['email', 'profile']
+
+
+class ConfigurationError(Exception):
+    """Raised when OAuth2 settings are not adequately configured."""
+    pass
+
+
+class NoCredentialsError(Exception):
+    """Raised when attempting to use credentials that do not exist."""
+    pass
+
+
+class OAuth2(object):
+    """Flask extension for making OAuth 2.0 easier.
+
+    The extension includes views that handle the entire auth flow and a
+    @required decorator to automatically ensure that user credentials are
+    available.
+
+    To configure::
+
+        app = Flask(__name__)
+
+        app.config['SECRET_KEY'] = 'your-secret-key'
+        app.config['OAUTH2_CLIENT_SECRETS_JSON'] = 'client_secrets.json'
+
+        oauth2 = OAuth2(app)
+
+
+    To use::
+
+        @app.route('/needs_credentials')
+        @oauth2.required
+        def example():
+            # http is authorized with the user's credentials and can be used
+            # to make http calls.
+            http = oauth2.http()
+
+            # Or, you can access the credentials directly
+            credentials = oauth2.credentials
+
+
+        @app.route('/info')
+        @oauth2.required
+        def info():
+            return "Hello, {}".format(oauth2.email)
+
+    """
+
+    def __init__(self, app=None, *args, **kwargs):
+        self.app = app
+        if app is not None:
+            self.init_app(app, *args, **kwargs)
+
+    def init_app(self, app, scopes=None, client_secrets_file=None,
+                 client_id=None, client_secret=None, authorize_callback=None,
+                 logout_callback=None):
+        self.app = app
```

Comment:
It is, but it's needed here because the app can be initialized later if the user is using an application factory

```
oauth2 = OAuth2()

def app_factory():
    app = flask()
    oauth2.init_app(app)
    return app
```